### PR TITLE
Add test to verify that word boundaries are still needed for whitespace-tokenized languages

### LINKED ISSUE #7167: Check if regex boundary is needed
**Description of Problem**:
Currently, we are adding a `\b` (boundary) to every regex for the `RegexFeaturizer` and `RegexEntityExtractor`. In https://github.com/RasaHQ/rasa/issues/7091 a user mentioned that this is not working for non-whitespace tokenizable language, such as Chinese. 
The question is, do we actually need to add the boundary at all? (See [slack thread](https://rasa-hq.slack.com/archives/CL3GUUUQ1/p1604394051233000) for reference.)

**Overview of the Solution**:
Figure out why the boundary was added in the first place and check if it is still needed. If not, remove it from the code.

### PR DESCRIPTION
**Pull request by koernerfelicia**
Fixes #7167

**Proposed changes**:
The tests verify that word boundaries are still needed for whitespace-tokenized languages. For example, if the pattern is `"how"`, we should match:
`"how are you"`
but **not**:
```
"take a shower"
"what a show"
"the wolf howled"
```

If users do wish to ignore word boundaries (i.e. match all of the sentences above), they can do so by setting `"use_word_boundaries"=False`

**Status (please check what you already did)**:
- [X] added some tests for the functionality
- [ ] updated the documentation N/A
- [X] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by koernerfelicia** [comment]
@samsucik I picked you randomly: I think this is a pretty easy code review that doesn't require much knowledge. Please let me know if I should find someone else!

**Review by samsucik — COMMENTED** [review]
The test looks good!

Since this was my first time working with this featuriser, there were many things I didn't understand -- hence my comments relating to improving the test clarity 🙂  Of course there's the tradeoff between being "compatible/consistent" with previously written code/tests vs being more verbose and/or understandable, but I personally lean towards the 2nd option. E.g. when I work on a bug fix and don't understand the code, together with writing the fix I also add comments that would help a future me.

File: tests/nlu/featurizers/test_regex_featurizer.py [hunk 1]
```diff
@@ -350,3 +350,50 @@ def test_regex_featurizer_case_sensitive(
     sequence_featrures, sentence_features = ftr._features_for_patterns(message, TEXT)
     assert np.allclose(sequence_featrures.toarray()[0], sequence_vector, atol=1e-10)
     assert np.allclose(sentence_features.toarray()[-1], sentence_vector, atol=1e-10)
+
+
+@pytest.mark.parametrize(
+    "sentence, expected, labeled_tokens, use_word_boundaries",
```
[the remark this answers is not in the document]
Comment (koernerfelicia) [reply tests/nlu/featurizers/test_regex_featurizer.py:356]: @samsucik do you think it's too aggressive to change all of them? I felt it was confusing at first too and then kind of forgot as I got used to it

[the remark this answers is not in the document]
Comment (samsucik) [reply tests/nlu/featurizers/test_regex_featurizer.py:356]: No, I don't think it's too aggressive. If we genuinely believe that one pattern is more clear than the other (and otherwise they're equivalent), then I would go for it. My only concern here is that in general once someone starts making changes aimed to improve code clarity, it can take too much of their time 🙂

File: tests/nlu/featurizers/test_regex_featurizer.py [hunk 2]
```diff
@@ -350,3 +350,50 @@ def test_regex_featurizer_case_sensitive(
...
+def test_regex_featurizer_with_and_without_boundaries(
+    sentence: Text,
+    expected: List[float],
+    labeled_tokens: List[float],
+    use_word_boundaries: bool,
+    spacy_nlp: Any,
+):
+    lookups = [
+        {"name": "how", "elements": ["how"],},
+    ]
```
[the remark this answers is not in the document]
Comment (koernerfelicia) [reply tests/nlu/featurizers/test_regex_featurizer.py]: That's a good point! I changed the test name and the changelog entry, and also added a comment in the test to explain why.

File: tests/nlu/featurizers/test_regex_featurizer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (koernerfelicia) [reply tests/nlu/featurizers/test_regex_featurizer.py]: >take too much of their time

Your time now too as the code review got larger :S. I changed all the tests with this `expected` pattern. That way people won't re-use it for new tests in the future.

File: tests/nlu/featurizers/test_regex_featurizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (samsucik) [reply tests/nlu/featurizers/test_regex_featurizer.py]: Awesome!

**Review by samsucik — APPROVED** [review]
Looks good, thanks for the changes towards making all the test cases more consistent and understandable!
